miasm2/ir/ssa.py

```diff
@@ -0,0 +1,279 @@
+import miasm2.expression.expression as m2_expr
+from miasm2.expression.expression_dissector import ExprDissector
+from miasm2.ir.ir import AssignBlock, irbloc
+
+
+class SSA(object):
+    """
+    Generic class for static single assignment (SSA) transformation
+
+    Handling of
+    - variable generation
+    - variable renaming
+    - conversion of an IRA block into SSA
+
+    Variables will be renamed to <variable>.<index>, whereby the
+    index will be increased in every definition of <variable>.
+
+    Memory expressions are stateless. The addresses are in SSA form,
+    but memory aliasing will occur. For instance, if it holds
+    that RAX == RBX.0 + (-0x8) and
+
+    @64[RBX.0 + (-0x8)] = RDX
+    RCX.0 = @64[RAX],
+
+    then it cannot be tracked that RCX.0 == RDX.
+    """
+
+    def __init__(self, ira):
+        """
+        Initialises generic class for SSA
+        :param ira: instance of IRA
+        """
+        # IRA instance
+        self.ira = ira
+
+        # SSA blocks
+        self.blocks = dict()
+
+        # stack for RHS
+        self._stack_rhs = dict()
+        # stack for LHS
+        self._stack_lhs = dict()
+
+        # dict of SSA expressions
+        self.expressions = dict()
+
+        # architecture variables
+        regs = self.ira.arch.regs.all_regs_ids
+        irdst = self.ira.IRDst
+        # class for parsing expressions
+        self._expr_dissect = ExprDissector(regs, irdst)
+
+    def transform(self, *args, **kwargs):
+        """Transforms into SSA"""
+        raise NotImplementedError("")
+
+    def get_block(self, block):
```

Comment:
It could be misleading to use `block` for an `asm_label` instance. In others code snippets, we preferably use `label`